sekoia_automation/checkpoint.py

```diff
@@ -0,0 +1,213 @@
+
+            if self._lock:
+                self._lock.release()
+
+            # if undefined, retrieve events from the {self._start_at} ago
+            if most_recent_date_seen_str is None:
+                self._most_recent_date_seen = (
+                    datetime.now(timezone.utc) - self._start_at
+                )
+                return self.from_datetime(self._most_recent_date_seen)
+
+            most_recent_date_seen = self.file_to_datetime(most_recent_date_seen_str)
+
+            if self._ignore_older_than:
+                # check if the date is older than the {self._ignore_older_than} ago
+                too_old = datetime.now(timezone.utc) - self._ignore_older_than
+                if most_recent_date_seen < too_old:
+                    most_recent_date_seen = too_old
+
+            self._most_recent_date_seen = most_recent_date_seen
+
+        return self.from_datetime(self._most_recent_date_seen)
+
+    @offset.setter
+    def offset(self, last_message_date: datetime) -> None:
+        if last_message_date is not None:
+            # convert to inner representation
+            last_message_date = self.to_datetime(last_message_date)
+
+            if self.offset is None or last_message_date > self.offset:
+                self._most_recent_date_seen = last_message_date
+
+                if self._lock:
+                    self._lock.acquire()
+
+                with self._context as cache:
+                    if self._subkey:
+                        if self._subkey not in cache:
+                            cache[self._subkey] = {}
+
+                        cache[self._subkey]["most_recent_date_seen"] = (
+                            self.datetime_to_file(self._most_recent_date_seen)
+                        )
+
+                    else:
+                        cache["most_recent_date_seen"] = self.datetime_to_file(
+                            self._most_recent_date_seen
+                        )
+
+                if self._lock:
+                    self._lock.release()
+
+
+class CheckpointDatetime(CheckpointDatetimeBase):
+    def from_datetime(self, dt):
+        return dt
+
+    def to_datetime(self, rp):
+        return rp
+
+
+class CheckpointTimestampMilliseconds(CheckpointDatetime):
+    def from_datetime(self, dt: datetime) -> int:
+        # inner representation -> timestamp [ms]
+        return round(dt.timestamp() * 1000.0)
+
+    def to_datetime(self, rp: int) -> datetime:
+        # timestamp [ms] -> inner representation
+        return datetime.fromtimestamp(rp).astimezone(timezone.utc)
+
+
+class CheckpointTimestampSeconds(CheckpointDatetime):
+    def from_datetime(self, dt: datetime) -> int:
+        # inner representation -> timestamp [ms]
+        return round(dt.timestamp())
+
+    def to_datetime(self, rp: int) -> datetime:
+        # timestamp [s] -> inner representation
+        return datetime.fromtimestamp(rp).astimezone(timezone.utc)
```

Comment:
Can we group these two classes and add a factor in the constructor of the mutual class to specify if the timestamp is in seconds, milliseconds or nanoseconds?